File: packages/flutter_shared/lib/tools/anniversary_tool.dart
/// 纪念日/倒数日工具
class AnniversaryTool {
  AnniversaryTool._();

  static ({int days, String description}) calculate(DateTime target) {
    final now = DateTime.now();
    final diff = target.difference(now);

    if (diff.isNegative) {
      final abs = diff.abs();
      final d = abs.inDays;
      if (d == 0) return (days: 0, description: '就是今天！🎉');
      if (d < 365) {
        return (days: -d, description: '已经过去了 $d 天');
      }
      final years = d ~/ 365;
      final remain = d % 365;
      if (remain == 0) {
        return (days: -d, description: '$years 周年纪念 🎂');
      }
      return (days: -d, description: '$years 年零 $remain 天');
    }

    final d = diff.inDays;
    if (d == 0) return (days: 0, description: '就是今天！🎉');
    if (d < 365) {
      return (days: d, description: '还有 $d 天');
    }
    final years = d ~/ 365;
    return (days: d, description: '还有 $d 天（约 $years 年）');
  }

  static String formatDate(DateTime d) {
    return '${d.year}年${d.month}月${d.day}日';
  }
}

File: miniapp/src/utils/tools/anniversaryTool.ts
/**
 * 纪念日/倒数日工具
 * 移植自 flutter_shared/tools/anniversary_tool.dart
 */

export interface AnniversaryResult {
  days: number;
  description: string;
}

export const AnniversaryTool = {
  calculate(target: Date): AnniversaryResult {
    const now = new Date();
    const diff = target.getTime() - now.getTime();

    if (diff < 0) {
      const abs = Math.abs(diff);
      const d = Math.floor(abs / (1000 * 60 * 60 * 24));
      if (d === 0) return { days: 0, description: '就是今天！🎉' };
      if (d < 365) return { days: -d, description: `已经过去了 ${d} 天` };
      const years = Math.floor(d / 365);
      const remain = d % 365;
      if (remain === 0) return { days: -d, description: `${years} 周年纪念 🎂` };
      return { days: -d, description: `${years} 年零 ${remain} 天` };
    }

    const d = Math.floor(diff / (1000 * 60 * 60 * 24));
    if (d === 0) return { days: 0, description: '就是今天！🎉' };
    if (d < 365) return { days: d, description: `还有 ${d} 天` };
    const years = Math.floor(d / 365);
    return { days: d, description: `还有 ${d} 天（约 ${years} 年）` };
  },

  formatDate(d: Date): string {
    return `${d.getFullYear()}年${d.getMonth() + 1}月${d.getDate()}日`;
  },
};
